### DISCUSSION TITLE
Why is Flax 7x slower than Haiku?

### DISCUSSION THREAD

**Discussion by nalzok** [google/flax#2109] [Q&A] [answered]
Here is a minimal, complete, and verifiable example: [mcve.zip](https://github.com/google/flax/files/8642517/mcve.zip). You can execute the scripts by

```bash
pipenv install
pipenv run python3 mcve_haiku.py
pipenv run python3 mcve_flax.py
```

Here is the background: I was trying to convert my research code from Haiku to Flax, but the performance got significantly worse for some reason. Basically `mcve_flax.py` and `mcve_haiku.py` perform the same training procedure to a linear neural network (at least that's what I intended), but the former takes 07:23 on my 2017 MacBook Pro, whereas the latter merely takes 01:05.

Am I doing anything wrong, or is this a known issue?

**Comment by marcvanzee** [comment, marked as answer]
This is your training loop:

```py
for X, y in (pbar := tqdm(itertools.islice(input_dataset, epochs), disable=quiet)):
        Hhat = LNN(neurons).apply({'params': state.params}, I)
        shat = np.linalg.svd(Hhat, compute_uv=False)

        if not quiet:
            error_s = np.array2string(
                s - shat,
                precision=2,
                separator=", ",
                formatter={"float_kind": lambda x: "%5.2f" % x},
            )
            error_H = f"error_H={jnp.linalg.norm(H-Hhat):6.3f}"
            pbar.set_description(f"{error_s}, {error_H}")

        state = train_step(tuple(neurons), state, X, y)
```

Each step contains two forward passes: one on line 2 and one inside `train_step`. Most problematically, the first call is not jitted, making it superslow.

Instead, you should only have one forward pass (like you have in Haiku), and get the statistics you need in this forward pass. 

Also, @melissatan recently added a `Sequential` layer so you can use that in the Flax module as well (just like you do in the Haiku module): https://github.com/google/flax/blob/main/flax/linen/combinators.py

**Reply by nalzok** [reply]
Thanks, I think the major problem is that I didn't JIT the first forward pass. In fact, the Haiku version also contains two forward passes: the first with `Hhat = forward_t.apply(params, I)` and the second with `grads = jax.grad(loss_fn_t.apply)(params, u, v)`. I cannot combine these two passes into one since they apply to different data (i.e. `I` and `u` respectively).

After using `jax.jit`, the Flax version finished in merely 6 seconds, which is over 10x faster compared Haiku! Care to share what makes Flax so performant? :)

**Reply by marcvanzee** [reply]
Awesome! I am not sure why Flax is so much faster, I believe in principle Flax and Haiku's performance on this example should be comparable.

Pinging @lenamartens from Haiku, is there something wrong in @nalzok's Haiku code?

**Reply by LenaMartens** [reply]
Hi, thanks Marc for the ping!

I was able to reproduce the performance difference between Flax and Haiku by jitting the `forward.apply` in the Flax code, but I noticed you're not jitting any of the functions in the Haiku code. After jitting as much as possible in both examples, I get very similar performance (around 700 iterations per second on both examples with a GPU backend, let me know if you see something different)

To clarify, I've restructured the `train` function in the Haiku example to the below, jitting as much as possible:
```
def train(ground_truth, neurons, epochs, quiet) -> tuple[float, list[np.ndarray]]:
    p = neurons[0]
    I = np.eye(p)

    np_rng = np.random.default_rng(42)
    u = np_rng.normal(size=p)
    H = ground_truth(u)

    batch_size = 256
    input_dataset = data_gen(np_rng, p, H, batch_size)

    s = np.linalg.svd(H, compute_uv=False)

    def forward(input):
        lnn = LNN(neurons)
        output = lnn(input)
        return output

    forward_t = hk.transform(forward)
    forward_t = hk.without_apply_rng(forward_t)

    def loss_fn(input, ground_truth):
        output = forward(input)
        return jnp.linalg.norm(output - ground_truth)

    loss_fn_t = hk.transform(loss_fn)
    loss_fn_t = hk.without_apply_rng(loss_fn_t)
    
    @jax.jit  # <!> new jit <!> 
    def update_fn(params, u, v):
      grads = jax.grad(loss_fn_t.apply)(params, u, v)
      params = jax.tree_util.tree_map(update_rule, params, grads)
      return params
    
    dummy_u, dummy_v = next(input_dataset)
    rng = jax.random.PRNGKey(42)
    params = loss_fn_t.init(rng, dummy_u, dummy_v)
    jitted_forward = jax.jit(forward_t.apply)  # <!> new jit <!>

    for u, v in (pbar := tqdm(islice(input_dataset, epochs), disable=quiet)):
        Hhat = jitted_forward(params, I)
        _, shat, _ = np.linalg.svd(Hhat, full_matrices=False)

        if not quiet:
            error_s = np.array2string(
                s - shat,
                precision=2,
                separator=", ",
                formatter={"float_kind": lambda x: "%5.2f" % x},
            )
            error_H = f"error_H={jnp.linalg.norm(H-Hhat):6.3f}"
            pbar.set_description(f"{error_s}, {error_H}")
        params = update_fn(params, u, v)
  

    Hhat = jitted_forward(params, I)
    loss = jnp.linalg.norm(H - Hhat)
    weights = list(np.array(params[f"lnn/~/linear_{i}"]["w"]) for i in range(3))

    return loss, weights
```

There's probably a more neat way to structure this, but the main additions are a `jit` around the `forward.apply` call, and a jitted `update_fn` (which includes `jax.grad(loss_fn_t.appy)` and the `tree_map` update)

**Reply by nalzok** [reply]
Nice catch! One more question: is there any mechanism to find all missed `jit` opportunities? I guess this is a common mistake by beginners, and it's really hard to debug such performance issues since the numerical result is correct.

**Reply by LenaMartens** [reply]
I'm afraid there's no such mechanism, and it's more of a best practice: you should put the `jit` at the outermost possible level, capturing as much jnp+python code as possible. Python overheads can quickly out-grow a compiled XLA computation, so if you for example move the `tree_map` into the jitted function, this can already make quite an improvement (even though intuitively, it does not represent much of the computation in your iteration).

Some code cannot be jitted, so should not be in a jit. In your code for example, you couldn't jit the `tqdm` updates (side-effecting operation). The trick lies in finding the largest jit boundary which is possible. (Some better explanation of this in JAX tutorials https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html#why-can-t-we-just-jit-everything)

In general, we find the following pattern is followed by many of our code-bases (ie. where the `jit` boundary usually lies):
```
for step in range(steps):  # your training loop
  params = jitted_update_fn(params, ....) # jitted update step
  if step % 10 == 0: # every so often, do some logging/saving actions
    ...
```
